## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.preprocessing import StandardScaler, OneHotEncoder
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras import layers
from sklearn.model_selection import train_test_split

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

2025-04-09 22:26:26.515710: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [3]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
y_df.head()

Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [4]:
# Create a list of at least 10 column names to use as X data
X_data = ['Education', 'Age', 'DistanceFromHome', 'JobSatisfaction','OverTime',
          'StockOptionLevel', 'WorkLifeBalance', 'YearsAtCompany','YearsSinceLastPromotion',
            'NumCompaniesWorked', 'MaritalStatus' ]


# Create X_df using your selected columns
X_df = attrition_df[X_data]


# Show the data types for X_df
print (X_df.dtypes)
print (X_df.nunique())
X_df.head()

Education                   int64
Age                         int64
DistanceFromHome            int64
JobSatisfaction             int64
OverTime                   object
StockOptionLevel            int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsSinceLastPromotion     int64
NumCompaniesWorked          int64
MaritalStatus              object
dtype: object
Education                   5
Age                        43
DistanceFromHome           29
JobSatisfaction             4
OverTime                    2
StockOptionLevel            4
WorkLifeBalance             4
YearsAtCompany             37
YearsSinceLastPromotion    16
NumCompaniesWorked         10
MaritalStatus               3
dtype: int64


Education  Age  DistanceFromHome  JobSatisfaction OverTime  \
0          2   41                 1                4      Yes   
1          1   49                 8                2       No   
2          2   37                 2                3      Yes   
3          4   33                 3                3      Yes   
4          1   27                 2                2       No   

   StockOptionLevel  WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0                 0                1               6                        0   
1                 1                3              10                        1   
2                 0                3               0                        0   
3                 0                3               8                        3   
4                 1                3               2                        2   

   NumCompaniesWorked MaritalStatus  
0                   8        Single  
1                   1       Married  
2                   6        Single  
3                   1       Married  
4                   9       Married

In [5]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df)

In [6]:
# Convert your X data to numeric data types however you see fit

print(X_train['OverTime'].value_counts())

OverTime
No     797
Yes    305
Name: count, dtype: int64


In [7]:
# Overtime & MaritalStatus are the only one to convert
# Convert Overtime simple mapping
X_train = pd.get_dummies(X_train, columns=['MaritalStatus'],prefix='MaritalStatus')
X_test = pd.get_dummies(X_test, columns=['MaritalStatus'],prefix='MaritalStatus')

X_train['OverTime'] = X_train['OverTime'].map({'Yes':1, 'No':0})
X_test['OverTime'] = X_test['OverTime'].map({'Yes':1, 'No':0})
print(X_train.dtypes)
print(X_test.dtypes)


print(X_train.dtypes)
print(X_test.dtypes)

Education                  int64
Age                        int64
DistanceFromHome           int64
JobSatisfaction            int64
OverTime                   int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
MaritalStatus_Divorced      bool
MaritalStatus_Married       bool
MaritalStatus_Single        bool
dtype: object
Education                  int64
Age                        int64
DistanceFromHome           int64
JobSatisfaction            int64
OverTime                   int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
MaritalStatus_Divorced      bool
MaritalStatus_Married       bool
MaritalStatus_Single        bool
dtype: object
Education                  int64
Age                        int64
DistanceFromHome           int64
JobSatisfaction

In [8]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler = scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled =X_scaler.transform(X_test)

In [9]:
# Create a OneHotEncoder for the Department column
dept_ohe = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
y_dept_ohe = dept_ohe.fit(y_train['Department'].values.reshape(-1,1))


# Create two new variables by applying the encoder
# to the training and testing data
y_train_dept_encoded = y_dept_ohe.transform(y_train['Department'].values.reshape(-1,1))
y_test_dept_encoded = y_dept_ohe.transform(y_test['Department'].values.reshape(-1,1))

y_train_dept_encoded

array([[0., 0., 1.],
       [0., 0., 1.],
       [0., 1., 0.],
       ...,
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [10]:
# Create a OneHotEncoder for the Attrition column
attr_ohe = OneHotEncoder(sparse_output=False)

# Fit the encoder to the training data
y_attr_ohe = attr_ohe.fit(y_train['Attrition'].values.reshape(-1,1))

# Create two new variables by applying the encoder
# to the training and testing data
y_train_attr_encoded = y_attr_ohe.transform(y_train['Attrition'].values.reshape(-1,1))
y_test_attr_encoded = y_attr_ohe.transform(y_test['Attrition'].values.reshape(-1,1))

y_train_attr_encoded

array([[1., 0.],
       [0., 1.],
       [1., 0.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

## Part 2: Create, Compile, and Train the Model

In [11]:
# Find the number of columns in the X training data.
cols = len(X_train.columns)

# Create the input layer
input_layer = tf.keras.layers.Input(shape=(cols,), name='input')

# Create at least two shared layers
shared_layer_1 = tf.keras.layers.Dense(units=64, activation='relu', name='Shared_Layer1')(input_layer)
shared_layer_2 = tf.keras.layers.Dense(units=128, activation='relu', name='Shared_Layer2')(shared_layer_1)

In [12]:
# Create a branch for Department
# with a hidden layer and an output layer


# Create the hidden layer
dept_layer = tf.keras.layers.Dense(units=32, activation='swish', name='Department_Hidden')(shared_layer_2)

# Create the output layer
dept_output_layer = tf.keras.layers.Dense(units=3, activation='softmax', name='Department_Output')(dept_layer)

In [13]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attr_layer = tf.keras.layers.Dense(units=32, activation='relu', name='Attrition_Hidden')(shared_layer_2)

# Create the output layer
attr_output_layer = tf.keras.layers.Dense(units=2, activation='sigmoid', name='Attrition_Output')(attr_layer)

In [14]:
# Create the model
model = tf.keras.Model(inputs=input_layer, outputs=[dept_output_layer, attr_output_layer])

# Compile the model
model.compile(optimizer='adam',
              loss={'Department_Output':'categorical_crossentropy',
                    'Attrition_Output' : 'binary_crossentropy'},
                    metrics={'Department_Output':'accuracy',
                             'Attrition_Output': 'accuracy'})

# Summarize the model
model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input (InputLayer)  │ (None, 13)        │          0 │ -                 │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Shared_Layer1       │ (None, 64)        │        896 │ input[0][0]       │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Shared_Layer2       │ (None, 128)       │      8,320 │ Shared_Layer1[0]… │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Department_Hidden   │ (None, 32)        │      4,128 │ Shared_Layer2[0]… │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Attrition_Hidden    │ (None, 32)        │      4,128 │ Shared_Layer2[0]… │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Department_Output   │ (None, 3)         │         99 │ Department_Hidde… │
│ (Dense)             │                   │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Attrition_Output    │ (None, 2)         │         66 │ Attrition_Hidden… │
│ (Dense)             │                   │            │                   │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 17,637 (68.89 KB)

 Trainable params: 17,637 (68.89 KB)

 Non-trainable params: 0 (0.00 B)

In [15]:
# Train the model
history = model.fit(X_train_scaled, {'Department_Output': y_train_dept_encoded,
                                     'Attrition_Output' : y_train_attr_encoded},
                                     epochs=10)

Epoch 1/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - Attrition_Output_accuracy: 0.6672 - Attrition_Output_loss: 0.6090 - Department_Output_accuracy: 0.6232 - Department_Output_loss: 0.8651 - loss: 1.4741
Epoch 2/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - Attrition_Output_accuracy: 0.8340 - Attrition_Output_loss: 0.4354 - Department_Output_accuracy: 0.6443 - Department_Output_loss: 0.7836 - loss: 1.2191
Epoch 3/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - Attrition_Output_accuracy: 0.8396 - Attrition_Output_loss: 0.4096 - Department_Output_accuracy: 0.6911 - Department_Output_loss: 0.7158 - loss: 1.1256
Epoch 4/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - Attrition_Output_accuracy: 0.8493 - Attrition_Output_loss: 0.3859 - Department_Output_accuracy: 0.6454 - Department_Output_loss: 0.7849 - loss: 1.1709
Epoch 5/10
35/35 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - Attrition_Output_accuracy: 0.8856 - Attrition_Output_loss: 0.3134 - Department_Output_accuracy: 0.6706 - Department_Output_loss: 0.7196 

In [16]:
# Evaluate the model with the testing data
eval_metrics = model.evaluate(X_test_scaled, 
                              {'Department_Output': y_test_dept_encoded,
                               'Attrition_Output': y_test_attr_encoded},
                               verbose=2)

print(f'[{eval_metrics[0]}, {eval_metrics[3]}, {eval_metrics[4]:.1}]')

12/12 - 0s - 24ms/step - Attrition_Output_accuracy: 0.8587 - Attrition_Output_loss: 0.3610 - Department_Output_accuracy: 0.6250 - Department_Output_loss: 0.8364 - loss: 1.2126
[1.2126342058181763, 0.8586956262588501, 0.6]


In [17]:
# Print the accuracy for both department and attrition
print(f'Attrition prediction accuracy: {eval_metrics[3]}')
print(f'Department prediction accuracy: {eval_metrics[4]:.1}')

Attrition prediction accuracy: 0.8586956262588501
Department prediction accuracy: 0.6


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1.Accuracy can be misleading for imbalanced tasks like attrition (few leavers) or department prediction (uneven department sizes).

Attrition: The ~86% accuracy might be inflated by the majority class. Possible Better metrics are precision, recall, F1-score, and AUC to assess the model's ability to identify employees seeking other opportunties outside of the current organization.

Department: The 60% accuracy needs context based on the number of departments and their sizes. Precision, recall, and F1-score (macro/weighted) per department, along with a confusion matrix, offer a clearer picture of classification performance

2. For the Attrition predictions output layer, I used 'swish' activation. While 'sigmoid' is commonly used for binary classification to produce a probability between 0 and 1, 'swish' is a more modern activation function that has shown promising results in some cases.

For the Department predictions output layer, I used 'softmax' activation. This is appropriate for multi-class classification, as it outputs a probability distribution across all possible departments, ensuring the probabilities sum to 1

3. Here are a couple of ways that this model could be improved:
    a. Feature Engineering: Derive new features from existing data that may improve predictions for attrition or department.
    b. Data Augmentation: If possible, increase the amount of training data.
    c. Hyperparameter Tuning: Optimize the model's architecture by experimenting with testing different activation functions, and varying the number of units in shared and hidden layers.